src/ua_securechannel.c

```diff
@@ -230,189 +248,160 @@ void UA_SecureChannel_revolveTokens(UA_SecureChannel* channel) {
-        ci->errorCode = UA_STATUSCODE_BADRESPONSETOOLARGE;
-    if(++ci->chunksSoFar > connection->remoteConf.maxChunkCount &&
-       connection->remoteConf.maxChunkCount > 0)
-        ci->errorCode = UA_STATUSCODE_BADRESPONSETOOLARGE;
+    /* The total message length */
+    UA_UInt32 pre_sig_length = (UA_UInt32)((uintptr_t)buf_body_end - (uintptr_t)ci->messageBuffer.data);
+    UA_UInt32 total_length = pre_sig_length;
+    if(channel->securityMode == UA_MESSAGESECURITYMODE_SIGN ||
+       channel->securityMode == UA_MESSAGESECURITYMODE_SIGNANDENCRYPT)
+        total_length +=
+            securityPolicy->symmetricModule.signingModule.signatureSize;
 
-    /* Prepare the chunk headers */
+    /* Encode the chunk headers at the beginning of the buffer */
+    UA_Byte *header_pos = ci->messageBuffer.data;
     UA_SecureConversationMessageHeader respHeader;
     respHeader.secureChannelId = channel->securityToken.channelId;
     respHeader.messageHeader.messageTypeAndChunkType = ci->messageType;
+    respHeader.messageHeader.messageSize = total_length;
     if(ci->errorCode == UA_STATUSCODE_GOOD) {
         if(ci->final)
             respHeader.messageHeader.messageTypeAndChunkType += UA_CHUNKTYPE_FINAL;
         else
             respHeader.messageHeader.messageTypeAndChunkType += UA_CHUNKTYPE_INTERMEDIATE;
-    }
-    else {
-        /* abort message */
-        ci->final = true; /* mark as finished */
+    } else {
         respHeader.messageHeader.messageTypeAndChunkType += UA_CHUNKTYPE_ABORT;
-        UA_String errorMsg;
-        UA_String_init(&errorMsg);
-        offset = UA_SECURE_MESSAGE_HEADER_LENGTH;
-        UA_UInt32_encodeBinary(&ci->errorCode, dst, &offset);
-        UA_String_encodeBinary(&errorMsg, dst, &offset);
-
-        bytesToWrite = offset -
-            UA_SECUREMH_AND_SYMALGH_LENGTH +
-            securityPolicy->symmetricModule.signingModule.signatureSize +
-            1; // + 1 for padding byte
-
-        // Pad the error message
-        if(ci->channel->securityMode == UA_MESSAGESECURITYMODE_SIGNANDENCRYPT) {
-            UA_Byte paddingSize = 0;
-            UA_Byte extraPaddingSize = 0;
-            securityPolicy->symmetricModule.calculatePadding(securityPolicy, bytesToWrite, &paddingSize, &extraPaddingSize);
-            UA_UInt16 totalPaddingSize = (UA_UInt16)((extraPaddingSize << 8) | paddingSize);
-
-            for(UA_UInt16 i = 0; i <= totalPaddingSize; ++i) // This is <= because the paddingSize byte also has to be written.
-            {
-                dst->data[offset] = paddingSize;
-                ++offset;
-            }
-
-            if(extraPaddingSize > 0) {
-                dst->data[offset] = extraPaddingSize;
-                ++offset;
-            }
-        }
     }
-    respHeader.messageHeader.messageSize = (UA_UInt32)offset;
+    UA_encodeBinary(&respHeader, &UA_TRANSPORT[UA_TRANSPORT_SECURECONVERSATIONMESSAGEHEADER],
+                    &header_pos, buf_end, NULL, NULL);
```

Comment:
Can't we pass the beginning of the body here instead of buf_end?
Just to make sure that we don't overwrite body content.